modules/miner/miner.go

```diff
@@ -194,10 +199,15 @@ func New(cs modules.ConsensusSet, tpool modules.TransactionPool, w modules.Walle
 // Close terminates all ongoing processes involving the miner, enabling garbage
 // collection.
 func (m *Miner) Close() error {
+	if err := m.tg.Stop(); err != nil {
```

Comment:
You need to check `tg.IsStopped()` every iteration of the loop, otherwise there's a deadlock.

Easy way to see is to write a test that turns on the mining, then calls `miner.Close()`. `miner.Close()` will never complete, at least as far as I can tell, it'll just hang. (you should write this test regardless)

Inside the loop, there's a place where it checks the value of `miningOn` and then returns. In that same place, you should add a check for `tg.IsStopped` as well.